pkgs/build-support/fetchgit/default.nix

```diff
@@ -84,7 +87,9 @@ stdenvNoCC.mkDerivation {
   # > from standard in as a newline-delimited list instead of from the arguments.
   sparseCheckout = builtins.concatStringsSep "\n" sparseCheckout;
 
-  inherit url rev leaveDotGit fetchLFS fetchSubmodules deepClone branchName nonConeMode postFetch;
+  inherit url rev fetchLFS fetchSubmodules deepClone branchName nonConeMode postFetch;
+
+  leaveDotGit = leaveDotGit || verifyCommit || verifyTag;
```

Comment:
Do you mean as part of the downloading fetcher?

Sadly not, as the fetcher is a fixed-output derivation.
If the verification would be part of it instead of a seperate component, it would not run if files with the same checksum already exists in the Nix store.
I tried to explain the problem in detail in https://landweber.xyz/ba.pdf Section 5.1.5
Also see https://github.com/NixOS/nixpkgs/pull/43233#issuecomment-589913599

From my understanding, enabling leaveDotGit should not inflate the source size badly, but I didn't test this.
In any case, this is only relevant when building the derivation. When substituting it, users will only downloaded the files (unless they themselves enabled leaveDotGit).

Reply:
> Sorry for the late reply, life was somewhat difficult this year. 😓

I'm sorry it was and thankful for your participation! :)

---

> It can in repositories with deep histories, like nixpkgs itself. (Sorry, I tried to produce an example but GitHub is having issues tonight, so fetchGit was taking ~forever.)

Usually `fetchgit` will only receive the latest commit (shallow clone).
Enabling deep cloning implies `leaveDotGit`, so the fetching would not add any storage overhead.

---

> In your BSc. thesis (and in particular the section you mentioned) the goal is to validate signatures when fetching nixpkgs itself, using builtins.fetchGit and without a known output hash. 

I'm sorry you got the impression, but the section "5.1.5 Verification in fixed-output derivations" does not talk about the builtin fetcher. It states:

> [T]here are two distinct ways of implementing verifying fetchers in Nix. [...] As almost all fetchers depend upon network access at build time for the file download, they typically contain a fixed-output derivation and thus require a hash. 

---

> If the output hash is known a priori, an attacker cannot manipulate[1](https://github.com/NixOS/nixpkgs/pull/330457#user-content-fn-h-3717bbc9915c3ca715983f004bd74f10) the fetcher's output.  However, git signature validation would still have value in nixpkgs to protect that initial fetch (by a maintainer) and authenticate the versions of upstream packages which are being pinned (by hash) in nixpkgs. In that case, wouldn't a single-pass implementation address the same threat model?

Right, that's a good question. Usually hash updates will not be performed by maintainers but by the merge bot.
Having the signature verification execute on every build ensures, that the bot updated to a signed release.
I also loath verifications that are not run every rebuild but sit in the code as if they were. It should not be possible to build a `fetchGit` derivation with the correct commit/sha256 hash but an invalid signature key.

---

> My concern is that the time- and storage-overhead of the two pass implementation, would cause users of fetchGit, whether nixpkgs contributors or nix users defining custom derivations, to simply not use signature validation (because it makes their builds slower or their system run out of disk space)

I share your concern regarding the time overhead (not space, as described above), because making the `.git` folder 'deterministic' is expensive on large repositories. However, as I don't see the one-pass approach as a viable option, I'd be willing to put this computation requirement on `fetchGit` users.